File: Atcoder/402/C_Dislike_Foods.java
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N >> M;

    // Read dishes
    vector<vector<int>> dishes(M);
    for (int i = 0; i < M; i++) {
        int K;
        cin >> K;
        vector<int> ingreds(K);
        for (int j = 0; j < K; j++) {
            cin >> ingreds[j];
        }
        dishes[i] = ingreds;
    }

    // Read B and build inverse array D[j] = day when ingredient j is overcome
    vector<int> D(N + 1, 0);
    for (int day = 1; day <= N; day++) {
        int ing;
        cin >> ing;
        D[ing] = day;
    }

    // cnt[t] = how many dishes become edible exactly on day t
    vector<int> cnt(N + 1, 0);
    for (const auto& ingreds : dishes) {
        if (ingreds.empty()) {
            // no ingredients → edible from day 0
            cnt[0]++;
        } else {
            int t = 0;
            for (int ing : ingreds) {
                t = max(t, D[ing]);
            }
            cnt[t]++;
        }
    }

    // build answer by prefix‐summing cnt[]
    vector<int> ans(N + 1, 0);
    int running = cnt[0];
    for (int day = 1; day <= N; day++) {
        running += cnt[day];
        ans[day] = running;
    }

    // output days 1..N
    for (int day = 1; day <= N; day++) {
        cout << ans[day] << '\n';
    }

    return 0;
}

File: Atcoder/402/C_Dislike_Foods.cpp
/**
*    created: 2025-04-19 17:37:08
**/
#include<bits/stdc++.h>
using namespace std;
#ifndef ONLINE_JUDGE
#include "../debug.cpp"
#else
#define debug(...)
#define debugArr(...)
#endif
#define int long long
#define show(x) cout << x << endl
#define sz(x) (int)(x.size())

signed main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL);
    auto solve_test = [&](int test_case){
        int n, m;
        cin >> n >> m;

        vector<vector<int>> ar(m);
        for (int i = 0; i < m; i++) {
            int K;
            cin >> K;
            ar[i].resize(K);
            for (int j = 0; j < K; j++) {
                cin >> ar[i][j];
            }

        }

        vector<int> days(n + 1);
        for (int i = 1; i <= n; i++) {
            int d;
            cin >> d;
            days[d] = i;
        }


        vector<int> cnt(n + 1, 0);
        for (const auto& x : ar) {
            if (x.empty()) {
                cnt[0]++;
            } 
            else {
                int t = 0;
                for (int ing : x) {
                    t = max(t, days[ing]);
                }
                cnt[t]++;
            }
        }


        vector<int> ans(n + 1, 0);
        int accu = cnt[0];
        for (int i = 1; i <= n; i++) {
            accu += cnt[i];
            ans[i] = accu;
        }

        for (int i = 1; i <= n; i++) {
            cout << ans[i] << '\n';
        }
    };
    int T = 1;
    //cin >> T;
    for(int i = 0; i < T; ++i){
        solve_test(i);
    }
}